site/blog/tutorials/iasql-on-ec2.mdx

```diff
@@ -0,0 +1,284 @@
+- `sudo apt-get upgrade -y`: used to install the newest versions of all packages currently installed on the system
+- `curl -fsSL https://get.docker.com -o get-docker.sh`: download the Docker install script, then it's executed by `sudo sh get-docker.sh`
+- `docker run -p 9876:9876 -p 5432:5432 --name iasql iasql/iasql`: run the IaSQL container
+Knowing the intent of each command, it's not hard to do the same functionality using `cloud-init`.
+```yaml
+#cloud-config
+package_update: true # apt-get update - this was in the `get-docker.sh` script so we didn't manually run it in our script
+package_upgrade: true # apt-get upgrade
+apt:
+  sources:
+    docker.list: # add the docker source for the package manager
+      source: deb [arch=amd64] https://download.docker.com/linux/ubuntu $RELEASE stable
+      keyid: 9DC858229FC7DD38854AE2D88D81803C0EBFCD88
+packages: # make sure docker is installed
+  - docker-ce
+  - docker-ce-cli
+runcmd: # simply run docker run command
+  - docker run -p 9876:9876 -p 5432:5432 --name iasql iasql/iasql
+```
+So why bother ourselves writing the above YAML file instead of simply using bash scripts? One advantage is that `cloud-init` directives translate to distro-compatible commands. We were using Ubuntu 22.04 as our EC2's OS (by selecting the corresponding AMI), therefore the `apt-get update` command would've worked on the server. But that's not always the case, and we might want our `user_data` script to work cross-distro without having to write complex if-else conditional statements in bash.
+
+:::note
+That if-else way is basically what the `get-docker.sh` script is doing:
+```bash
+# a part of get-docker.sh script - changed to make the statements easier to follow
+case "$lsb_dist" in
+  ubuntu|debian|raspbian)
+    # uses apt-get
+  centos|fedora|rhel)
+    # uses dnf or yum to install packages
+    if [ "$lsb_dist" = "fedora" ]; then
+      pkg_manager="dnf"
+    else
+      pkg_manager="yum"
+    fi
+  sles)
+    # uses zypper to install packages
+    $sh_c "zypper -q install -y $pkgs"
+  *)
+    # the script won't work on macos
+    if is_darwin; then
+      echo "ERROR: Unsupported operating system 'macOS'"
+      echo "Please get Docker Desktop from https://www.docker.com/products/docker-desktop"
+    fi
+esac
+```
+:::
+
+Good thing about cloud-init is that it's widely supported: you can use it on GCP, AWS, DigitalOcean, and many other cloud providers to provision your servers. Their documentation says:
+> `Cloud-init` is the industry standard multi-distribution method for cross-platform cloud instance initialisation. It is supported across all major public cloud providers, provisioning systems for private cloud infrastructure, and bare-metal installations.
```

Comment:
Not really a comment on this itself, but I consider this `cloud-init` yaml format misguided because you can't escape platform-specific information in it.

Reply:
I thought about this `cloud-init` part and I think if the commands aren't cross-distro enough, it won't make sense to have it in this post. I can come to it in some other post that we can actually leverage it to have something cool.